persistence/jpa/src/main/java/org/infinispan/persistence/jpa/JpaStore.java

```diff
@@ -36,25 +37,29 @@
 import org.infinispan.commons.configuration.ConfiguredBy;
 import org.infinispan.commons.marshall.StreamingMarshaller;
 import org.infinispan.commons.persistence.Store;
+import org.infinispan.commons.util.AbstractIterator;
 import org.infinispan.executors.ExecutorAllCompletionService;
-import org.infinispan.filter.KeyFilter;
 import org.infinispan.marshall.core.MarshalledEntry;
 import org.infinispan.marshall.core.MarshalledEntryFactory;
 import org.infinispan.metadata.InternalMetadata;
-import org.infinispan.persistence.TaskContextImpl;
 import org.infinispan.persistence.jpa.configuration.JpaStoreConfiguration;
 import org.infinispan.persistence.jpa.impl.EntityManagerFactoryRegistry;
 import org.infinispan.persistence.jpa.impl.MetadataEntity;
 import org.infinispan.persistence.jpa.impl.MetadataEntityKey;
 import org.infinispan.persistence.jpa.impl.Stats;
 import org.infinispan.persistence.spi.AdvancedLoadWriteStore;
 import org.infinispan.persistence.spi.InitializationContext;
+import org.infinispan.util.KeyValuePair;
 import org.infinispan.util.TimeService;
 import org.infinispan.util.logging.Log;
 import org.infinispan.util.logging.LogFactory;
 
+import io.reactivex.Flowable;
+import io.reactivex.schedulers.Schedulers;
+
 /**
- *
+ * NOTE: This store can return expired entries on any given operation if {@link JpaStoreConfiguration#storeMetadata()}
+ * was not set to true.
```

Comment:
Does that include `publishKeys`? It doesn't really return entries, but I don't see any additional checks.

Reply:
Actually I'd like you to change the javadoc of `publishKeys` in `AdvancedCacheLoader` and `PersistenceManager`, because they say "This publisher will never return a key which belongs to an expired entry"

Reply:
Hrmm well that is the contract, which I would like to keep on `AdvancedCacheLoader`. JpaStore is special in that it violates this contract, which is why I would rather put it here. I don't want an implementation to return expired keys. 

For this reason DummyStore doesn't implement publishKeys, because the metadata is stored with entry so we have to read that as well.

Reply:
I don't think a javadoc here is enough, because there's nothing anyone can do about `JpaStore` breaking the contract:
* We won't add a special case for `JpaStore` in core, we're going to call `publishKeys` based on the `AdvancedCacheLoader` javadoc.
* Users call methods on the cache, they don't necessarily know which cache method calls which store method internally.

IMO we should throw an exception when the user tries to store mortal entries in a `JpaStore` if we can't read the metadata back. It's a bit complicated by the async writer catching all the exceptions, but at least they'll see some errors in the log. And ideally I'd like to make expiration support optional, both at the store level and at the cache level, so we can throw that exception right from `CacheImpl`.

Reply:
Hrmm I think it might be easier to log another JIRA for this.

It would cover

1. I would like to throw an exception at start up if JPAStore is configured not being able to store metadata. Unfortunately we can't do this, since they could be using the store as a read only.
2. We probably can only throw exceptions at runtime if an entry is written to the JPAStore with expiration.

Either way neither of these are great options, thus why I would recommend a separate JIRA to talk about this.

Reply:
I created https://issues.jboss.org/browse/ISPN-9143 to continue this.